pkgs/tools/nix/nixos-option/nixos-option.cc

```diff
@@ -168,7 +170,13 @@ const std::string appendPath(const std::string & prefix, const std::string & suf
     return prefix + "." + quoteAttribute(suffix);
 }
 
-bool forbiddenRecursionName(std::string name) { return (!name.empty() && name[0] == '_') || name == "haskellPackages"; }
+bool forbiddenRecursionName(const nix::Symbol symbol, const nix::SymbolTable & symbolTable) {
+    // note: this is created from a pointer
+    // According to standard, it may never point to null, and hence attempts to check against nullptr are not allowed.
+    // However, at the time of writing, I am not certain about the full implications of the omission of a nullptr check here.
+    const std::string & name = symbolTable[symbol];
+    return (!name.empty() && name[0] == '_') || name == "haskellPackages";
```

Comment:
can you add a TODO for this list of ad-hoc "forbidden" recursion packages?

Reply:
I am not fully sure what I should write into that tbh. Maybe something like this?

```suggestion
    # TODO: figure out why haskellPackages is not recursed here
    return (!name.empty() && name[0] == '_') || name == "haskellPackages";
```

Reply:
It was added with the rewrite in 59c5bfc86b75247cb48539eeaaea2a3c5f320b1d without explenation.